Complete and refactor web worker chapter (#22)
NLESC-JCER/cpp2wasm: README.md

Executing a long running C++ method will block the browser from running any other code like updating the user interface. In order to avoid this, the method can be run in the background using [web workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers). A web worker runs in its own thread and can be interacted with from JavaScript using messages.

> We need to instantiate a web worker which will implement later in `src/js/worker.js`.

Reviewer: ```suggestion
We need to instantiate a web worker which we will implement later in `src/js/worker.js`.
```

```{.js #worker-consumer}
// this JavaScript snippet is later referred to as worker-consumer
const worker = new Worker('worker.js');
```

We need to send the worker a message with description for the work it should do.

> ```{.js #worker-consumer}

Reviewer: add a comment line here

worker.postMessage({
  type: 'CALCULATE',
  data: { epsilon: 0.001, guess: -20 }
});
```

In the web worker we need to listen for incoming messages.

```{.js #worker-provider-onmessage}
// this JavaScript snippet is later referred to as worker-provider-onmessage
onmessage = function(message) {
  <<handle-message>>
};
```

Before we can handle the message we need to import the Web Assembly module.

```{.js file=src/js/worker.js}
// this JavaScript snippet is stored as src/js/worker.js
importScripts('newtonraphsonwasm.js');

<<worker-provider-onmessage>>
```

> We can only handle the `CALCULATE` message when the Web Assembly module is loaded and initialized.

Reviewer: ```suggestion
We can handle the `CALCULATE` message only after the Web Assembly module is loaded and initialized.
```

// this JavaScript snippet is later referred to as worker-consumer
const worker = new Worker('worker.js');
```

We need to send the worker a message with description for the work it should do.

```{.js #worker-consumer}
// this JavaScript snippet is appended to worker-consumer
worker.postMessage({
  type: 'CALCULATE',
  data: { epsilon: 0.001, guess: -20 }
});
```

In the web worker we need to listen for incoming messages.

```{.js #worker-provider-onmessage}
// this JavaScript snippet is later referred to as worker-provider-onmessage
onmessage = function(message) {
  <<handle-message>>
};
```

Before we can handle the message we need to import the Web Assembly module.

```{.js file=src/js/worker.js}
// this JavaScript snippet is stored as src/js/worker.js
importScripts('newtonraphsonwasm.js');

<<worker-provider-onmessage>>
```

We can handle the `CALCULATE` message only after the Web Assembly module is loaded and initialized.

```{.js #handle-message}
// this JavaScript snippet is before referred to as handle-message
if (message.data.type === 'CALCULATE') {
  createModule().then((module) => {
    <<perform-calc-in-worker>>
    <<post-result>>
  });
}
```

> Let's find the root based on the data parameters in the incoming message.

Reviewer: ```suggestion
Let's calculate the result (root) based on the data parameters in the incoming message.
```

```{.js #perform-calc-in-worker}
// this JavaScript snippet is before referred to as <<perform-calc-in-worker>>
const epsilon = message.data.payload.epsilon;
const finder = new module.NewtonRaphson(epsilon);
const guess = message.data.payload.guess;
const root = finder.find(guess);
```

And send the result back to the web worker consumer as a outgoing message.

```{.js #post-result}
// this JavaScript snippet is before referred to as <<post-result>>
postMessage({
  type: 'RESULT',
  payload: {
    root: root
  }
});
```

Listen for messages from worker and when a result message is received put the result in the HTML page like we did before.

```{.js #worker-consumer}
// this JavaScript snippet is appended to <<worker-consumer>>
worker.onmessage = function(message) {
  if (message.data.type === 'RESULT') {
    const root = message.data.payload.root;
    <<render-answer>>
  }
}
```

> Like before we need a HTML page to run the JavaScript, but now we don't need to import the `newtonraphsonwasm.js` file here as this is done in the `worker.js` file.

Reviewer: ```suggestion
Like before we need a HTML page to run the JavaScript, but now we don't need to import the `newtonraphsonwasm.js` file here as it is imported in the `worker.js` file.
```

```{.html file=src/js/example-web-worker.html}
<!doctype html>
<!-- this HTML page is stored as src/js/example-web-worker.html -->
<html>
  <script>
    <<worker-consumer>>
  </script>
</html>
```

Like before we also need to host the files in a web server with

```shell
python3 -m http.server 8000
```

> Visit [http://localhost:8000/src/js/example-web-worker.html](http://localhost:8000/src/js/example-web-worker.html) to see the root answer.

Reviewer: ```suggestion
Visit [http://localhost:8000/src/js/example-web-worker.html](http://localhost:8000/src/js/example-web-worker.html) to see the result of the calculation.
```

> The root finding answer was calculated using the C++ algorithm compiled to a WebAssembly module, imported in a web worker (seperate thread), executed by JavaScript with mesages to/from the web worker and rendered on a HTML page.

Reviewer: ```suggestion
The result of root finding was calculated using the C++ algorithm compiled to a WebAssembly module, imported in a web worker (separate thread), executed by JavaScript with message to/from the web worker and rendered on a HTML page.
```
